### PR TITLE
[RLlib] Add DTTorchModel

### PR DESCRIPTION
**Pull request by charlesjsun**
## Why are these changes needed?

Part of adding DT to RLlib.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/dt/tests/test_dt_model.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+            gym.spaces.Discrete(action_dim),
+        ]
+
+        B, T = 2, 3
+
+        for action_space in action_spaces:
+            # Generate input dict.
+            input_dict = SampleBatch(
+                {
+                    SampleBatch.OBS: np.arange(
```
Comment (avnishn) [inline rllib/algorithms/dt/tests/test_dt_model.py:162]: store in variable for clarity when reusing this thx :)

**Review by avnishn — COMMENTED** [review]
had some minor nits otherwise lgtm

File: rllib/algorithms/dt/dt_torch_model.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+    def get_prediction(
+        self,
+        model_out: TensorType,
+        input_dict: SampleBatch,
+    ) -> Dict[str, TensorType]:
+        """Computes the output of a forward pass of the decision transformer.
+
+        Args:
+            model_out: output observation tensor from the base model, [B x T x obs_dim]
```
Comment (kouroshHakha) [inline rllib/algorithms/dt/dt_torch_model.py:143]: for shape use [B,T,obs_dim] to avoid confusion with [BxTxobs_dim]

File: rllib/algorithms/dt/dt_torch_model.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                if model_config["use_obs_output"]
+                    OBS: [B x T x obs_dim] of predicted observations
+                if model_config["use_return_output"]
+                    RETURNS_to_GO: [B x T x 1] of predicted returns to go
+        """
+        B, T, *_ = model_out.shape
+
+        obs_embeds = self.obs_encoder(model_out)
+        actions_embeds = self.action_encoder(input_dict[SampleBatch.ACTIONS])
+        # Note: rtg might have an extra element at the end for targets
```
Comment (kouroshHakha) [inline rllib/algorithms/dt/dt_torch_model.py:162]: Can you add a comment on why it "might" have T or T+1 elements? basically mention inference vs. training differences.

File: rllib/algorithms/dt/tests/test_dt_model.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+            for _ in range(10):
+                model = DTTorchModel(
+                    observation_space,
+                    action_space,
+                    num_outputs,
+                    model_config,
+                    "model",
+                )
+                # so dropout is not in effect
+                model.eval()
```
[the remark this answers is not in the document]
Comment (charlesjsun) [reply rllib/algorithms/dt/tests/test_dt_model.py:117]: done

File: rllib/algorithms/dt/tests/test_dt_model.py (hunk 1 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/tests/test_dt_model.py]: done

File: rllib/algorithms/dt/dt_torch_model.py (hunk 2 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/dt_torch_model.py]: done

File: rllib/algorithms/dt/dt_torch_model.py (hunk 3 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/dt_torch_model.py:162]: done

File: rllib/algorithms/dt/tests/test_dt_model.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        observation_space = gym.spaces.Box(-1.0, 1.0, shape=(num_outputs,))
+
+        action_dim = 5
+        action_spaces = [
+            gym.spaces.Box(-1.0, 1.0, shape=(action_dim,)),
+            gym.spaces.Discrete(action_dim),
+        ]
+
+        B, T = 3, 4
+
```
Comment (kouroshHakha) [inline rllib/algorithms/dt/tests/test_dt_model.py:101]: **TODO** maybe create the shared stuff inside setupclass method to avoid code repetition.

Comment (charlesjsun) [reply rllib/algorithms/dt/tests/test_dt_model.py:101]: I am actually changing the model_config just to make sure i didn't do any copypaste error where I used a constant. But I did move input_dict to its own thing

File: rllib/algorithms/dt/tests/test_dt_model.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
+import unittest
```
Comment (kouroshHakha) [inline rllib/algorithms/dt/tests/test_dt_model.py:1]: There is no test here that actually tests the causality. This is very important since the performance of the model relies on it. For this you need to apply non-zero input and print out the attention masks. There should be attention map should be upper triangular indicating there is no attention paid to future. Also you need to test how that attention map is applied to r_t, s_t, and a_t within a single time-step.

File: rllib/algorithms/dt/tests/test_dt_model.py (hunk 5 above)
Comment (kouroshHakha) [reply rllib/algorithms/dt/tests/test_dt_model.py:101]: no I meant the sample_batch that you create in each case.

**Comment by kouroshHakha** [comment]
Loved this PR. Very clean and clear. the only thing remaining is adding a unittest that checks for causality within the model.

File: rllib/algorithms/dt/tests/test_dt_model.py (hunk 5 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/tests/test_dt_model.py:101]: yes that's what I did, i'm not sure why it's not showing up when clicking "view changes", but if you go to the lastest commit you'll see the change in files

File: rllib/algorithms/dt/tests/test_dt_model.py (hunk 6 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/tests/test_dt_model.py:1]: done

**Review by kouroshHakha — APPROVED** [review]
@richardliaw LGTM. Please merge after the Segmentation buffer PR is merged.
